File: prettier-plugin-csharpier/src/Printer/Types/DelegateDeclaration.ts
import { Doc } from "prettier";
import { printConstraintClauses } from "../PrintConstraintClauses";
import { PrintMethod } from "../PrintMethod";
import {
    HasIdentifier,
    SyntaxToken,
    printIdentifier,
    printSyntaxToken,
    SyntaxTreeNode,
    printModifiers,
} from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { AttributeListNode } from "./AttributeList";
import { ParameterListNode } from "./ParameterList";
import { TypeParameterConstraintClauseNode } from "./TypeParameterConstraintClause";
import { TypeParameterListNode } from "./TypeParameterList";

export interface DelegateDeclarationNode extends SyntaxTreeNode<"DelegateDeclaration"> {
    attributeLists: AttributeListNode[]; // TODO
    modifiers: SyntaxToken[];
    delegateKeyword?: SyntaxToken;
    returnType?: SyntaxTreeNode;
    identifier: SyntaxToken;
    typeParameterList?: TypeParameterListNode;
    parameterList?: ParameterListNode;
    constraintClauses: TypeParameterConstraintClauseNode[];
    arity?: number;
}

export const printDelegateDeclaration: PrintMethod<DelegateDeclarationNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [];
    parts.push(printModifiers(node));
    parts.push(printSyntaxToken(node.delegateKeyword), " ");
    parts.push(path.call(print, "returnType"));
    parts.push(" ", printIdentifier(node));
    if (node.typeParameterList) {
        parts.push(path.call(print, "typeParameterList"));
    }
    parts.push(path.call(print, "parameterList"));
    printConstraintClauses(node, parts, path, options, print);
    parts.push(";");

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/DelegateDeclaration.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class DelegateDeclaration
{
    public static Doc Print(DelegateDeclarationSyntax node, FormattingContext context)
    {
        var docs = new List<Doc>
        {
            AttributeLists.Print(node, node.AttributeLists, context),
            Modifiers.PrintSorted(node.Modifiers, context),
            Token.PrintWithSuffix(node.DelegateKeyword, " ", context),
            Node.Print(node.ReturnType, context),
            { " ", Token.Print(node.Identifier, context) },
        };
        if (node.TypeParameterList != null)
        {
            docs.Add(Node.Print(node.TypeParameterList, context));
        }
        docs.Add(
            Node.Print(node.ParameterList, context),
            ConstraintClauses.Print(node.ConstraintClauses, context),
            Token.Print(node.SemicolonToken, context)
        );
        return Doc.Concat(docs);
    }
}
